### DISCUSSION TITLE
bug: same id for different network cause a bug in switching network

### DISCUSSION THREAD

**Discussion by dev-aly3n** [wevm/wagmi#2662] [Ideas]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Package Version

1.2

### Current Behavior

we have a test network on tenderly that is actually a fork of Arbitrum one. chainID and all other data of network are same except the rpc url. for now both wagmi and rainbowkit has the same behavior and they doesn't care if the network is a fork with same id

current behavior: `switchNetwork?.(network.id)`

### Expected Behavior

`switchNetwork` function must accept other optional arguments like this:

expected behavior: `switchNetwork?.(id, rpcUrl?, name?)` (it's not a breaking change)
or a better option: `switchNetwork?.({...allNetworkProperties})`

### Steps To Reproduce

you can fork a network here: [tenderly](tenderly.co)

### Anything else?

it's a common usecase to fork a network to run test.

**Comment by dev-aly3n** [comment]
the Contributing guide is not working. I followed the steps but it shows a lots of error. also playground not working (it not updated based on changes in the packages)

**Comment by dev-aly3n** [comment]
ok I just share my thoughts here:
(it's very silly and simple to just focus on the pronlem, and also I just follow up the Injected connector here)
inside connectors base file should be like this:
`switchChain?(chainId: number, rpcUrl?:string): Promise<Chain>`


then we can add another error inside connector injected `switchChain` method:

```ts
async switchChain(chainId: number, rpcUrl?: string) {
    const provider = await this.getProvider()
    if (!provider) throw new ConnectorNotFoundError()
    const id = numberToHex(chainId)
    const matchedChain = this.chains.find((x) => x.id === chainId)
    if (matchedChain && rpcUrl) throw new chainWithProvidedRpcUrlNotAddedError() // the error
...}
```

then we can add this condition in `catch` block to handle that error

```ts
   if (
        (error as Error)?.name === chainWithProvidedRpcUrlNotAddedError.name &&
        rpcUrl
      ) {
        try {
          await provider.request({
            method: 'wallet_addEthereumChain',
            params: [
              {
                chainId: id,
                chainName: chain.name,
                nativeCurrency: chain.nativeCurrency,
                rpcUrls: [rpcUrl], //added rpcUrl here
                blockExplorerUrls: this.getBlockExplorerUrls(chain),
              },
            ],
          })

          const currentChainId = await this.getChainId()
          if (currentChainId !== chainId)
            throw new UserRejectedRequestError(
              new Error('User rejected switch after adding network.'),
            )

          return chain
        } catch (error) {
          throw new UserRejectedRequestError(error as Error)
        }
      }
```

so now when we use `switchNetwork` function with second argument (rpcUrl) and if chainId matches an existing supported id, then it will throw that error and instead of `wallet_switchEthereumChain` method it calls `wallet_addEthereumChain`

I couldn't manage to test it as the workspace was full of error.

thanks in advance

**Comment by tmm** [comment]
Moving to feature request. As it stands right now, unlikely that we add support for this by using the RPC URL. Maybe in the future, you could pass a full `Chain` type.
